pkg/state/appender.go

```diff
@@ -429,30 +515,53 @@ func (a *txAppender) appendTx(tx proto.Transaction, params *appendTxParams) erro
 		// Invoke and Exchange transactions should be handled differently.
 		// They may fail, and will be saved to blockchain anyway.
 		fallibleInfo := &fallibleValidationParams{appendTxParams: params, senderScripted: accountHasVerifierScript, senderAddress: senderAddr}
-		applicationRes, err = a.handleFallible(tx, fallibleInfo)
+		applicationRes, err = a.handleInvokeOrExchangeTransaction(tx, fallibleInfo)
 		if err != nil {
-			msg := "fallible validation failed"
-			if txID, err2 := tx.GetID(a.settings.AddressSchemeCharacter); err2 == nil {
-				msg = fmt.Sprintf("fallible validation failed for transaction '%s'", base58.Encode(txID))
-			}
-			return errs.Extend(err, msg)
+			return errors.Errorf("failed to handle invoke or exchange transaction, %v", err)
 		}
 		// Exchange and Invoke balances are validated in UTX when acceptFailed is false.
 		// When acceptFailed is true, balances are validated inside handleFallible().
 		needToValidateBalanceDiff = params.validatingUtx && !params.acceptFailed
-	default:
-		// Execute transaction's scripts, check against state.
-		txScriptsRuns, err := a.checkTransactionScripts(tx, accountHasVerifierScript, params)
+	case proto.EthereumMetamaskTransaction:
+		ethTx, ok := tx.(*proto.EthereumTransaction)
+		if !ok {
+			return errors.New("failed to cast interface transaction to ethereum transaction structure")
+		}
+		err = ethTx.GenerateID(a.settings.AddressSchemeCharacter)
 		if err != nil {
-			return err
+			return errors.Errorf("failed to generate transaction id for ethereum transaction, %v", err)
+		}
+		ethTx.TxKind, err = a.guessEthereumTransactionKind(ethTx, params)
```

Comment:
The same question, possibly the eth transaction kind should be guessed earlier.

Reply:
I assume we can't guess ethereum transaction before type cast